fixed bytes operations in warplib

AlejandroLabourdette opened the pull request.

rodrodros committed:
Make warplib a scarb project

rodrodros committed:
Add maths module

rodrodros committed:
update corelib to alpha 7

rodrodros committed:
Add arm binaries from alpha 7

rodrodros committed:
Add warp plugin binaries

rodrodros committed:
Rework structure for

rodrodros committed:
Fix warp dependency

rodrodros committed:
Fix warplib and delete files to ignore

rodrodros committed:
delete autogenerted warplib scripts

rodrodros committed:
Add corelib and bin for linux_x64

rodrodros committed:
Scarb package name replaces - for _

rodrodros committed:
Update compileCairo1 to use new binary

rodrodros committed:
Update compilation testing framework

rodrodros committed:
Some comments

rodrodros committed:
Merge branch 'wm-plugin-integration' of https://github.com/NethermindEth/warp into wm-plugin-integration

rodrodros committed:
Address review

rodrodros committed:
Add unimplemented traits for WarpMemory

rodrodros committed:
Typo

rodrodros committed:
Merge branch 'warp-memory-base-for-cairo1' into warplib-memory-to-cairo1

rodrodros committed:
Delete maths.cairo

rodrodros committed:
Fix comp errors

rodrodros committed:
Merge branch 'cairo-1.0' into warplib-memory-to-cairo1

rodrodros committed:
Add array handling logic

rodrodros committed:
Fix some compilation errors

rodrodros committed:
Make memory operations safe

rodrodros committed:
Merge branch 'cairo-1.0' into warplib-memory-to-cairo1

rodrodros committed:
fix typo

AlejandroLabourdette committed:
functions for bytes

AlejandroLabourdette committed:
Merge branch 'warplib-memory-to-cairo1' into warp-memory-to-cairo1-alab

rodrodros commented on warplib/src/warp_memory/bytes.cairo:
```diff
@@ -1,6 +1,62 @@
 
 trait WarpMemoryBytesTrait{
     fn bytes_to_fixed_bytes32(ref self: WarpMemory, location: felt252) -> u256;
+    fn wm_bytes_to_fixed_bytesN(self: WarpMemory, bytesLoc: felt252, width: felt252) -> felt252 ;
```
We can now remove the `wm_` prefix, we no longer need it

rodrodros commented on warplib/src/warp_memory/bytes.cairo:
```diff
@@ -1,6 +1,62 @@
+impl WarpMemoryBytesImpl of WarpMemoryBytesTrait{
+    fn bytes_to_fixed_bytes32(self: WarpMemory, bytesLoc: felt252) -> u256 {
+        // TODO Is this self call to read_felt enough, don't we need to import from WarpMemoryReadTrait??
+        let data_len = self.read_u256(bytesLoc);
```
length from bytes is now stored as a felt instead of a u256. You should read a single felt from here

rodrodros commented on warplib/src/warp_memory/bytes.cairo:
```diff
@@ -1,6 +1,62 @@
 }
+
+impl WarpMemoryBytesImpl of WarpMemoryBytesTrait{
+    fn bytes_to_fixed_bytes32(self: WarpMemory, bytesLoc: felt252) -> u256 {
```
camelCase to snake_case

rodrodros commented on warplib/src/warp_memory/bytes.cairo:
```diff
@@ -1,6 +1,62 @@
+    }
+
+    // TODO This function will convert to fixed but store it in a felt, so will only be saved the last 31 bytes. Is this the approach we want??
+    fn wm_bytes_to_fixed_bytesN(self: WarpMemory, bytesLoc: felt252, width: felt252) -> felt252 {
```
camelCase to snake_case of `bytesLoc`

rodrodros commented on warplib/src/warp_memory/bytes.cairo:
```diff
@@ -1,6 +1,62 @@
+            return self._bytes_to_fixed_recursive(bytesDataLoc + 1, targetWidth - 1, data_len, 256 * representation);
+        } else {
+            // TODO Is this self call to read_felt enough, don't we need to import from WarpMemoryReadTrait??
+            let byte = self.read_felt(bytesDataLoc);
```
self.read_felt to self.read

AlejandroLabourdette replied:
Then why would be needed WarpMemoryReadTrait?

https://github.com/NethermindEth/warp/blob/cac9aa92e682b537b0d02426c8c93594889a3639/warplib/src/warp_memory/read.cairo#L4-L6

piwonskp replied:
See https://github.com/NethermindEth/warp/pull/1038#discussion_r1181965536

piwonskp committed:
Use Serde trait to store/retrieve from memory (#1040)

* Use Serde trait to store/retrieve from memory

* Fix missing import

* Fix function naming in the trait

* Replace individual memory reads with a call to warp_memory

* Divide MemoryTrait into low level trait and external interface

* Remove unused import

* Add missing newline

* Fix wrong imports

* Cell accessor -> Warp memory impl

* Fix missing imports

* Use memory's own `read`instead of Felt252Dict's method

* Add memory implicit to memoryRead

* Fix arrayLiteral implicits

---------

Co-authored-by: Piotr Piwoński <piwonskp>

piwonskp commented on warplib/src/warp_memory/read.cairo:
```diff
@@ -3,10 +3,17 @@ use warplib::warp_memory::WarpMemoryTrait;
+        let low = self.read(index);
+        let high = self.read(index + 1);
+        u256{low, high}
+    }
```
We should use MemoryAccessor's `retrieve` instead

AlejandroLabourdette committed:
remove wm prefix

AlejandroLabourdette committed:
Merge branch 'warplib-memory-to-cairo1' into warp-memory-to-cairo1-alab

rodrodros committed:
Address review

AlejandroLabourdette committed:
camel case to snake case

AlejandroLabourdette committed:
read bytes length as a felt

AlejandroLabourdette committed:
fixed compilation errors

rodrodros committed:
Update read_id to get_or_create_id

rodrodros committed:
Remove warplib/target

piwonskp commented:
Let's add some unit testing as I've suggested in https://github.com/NethermindEth/warp/pull/1038#issuecomment-1534522079

piwonskp commented on warplib/src/warp_memory/bytes.cairo:
```diff
@@ -1,6 +1,51 @@
+            } else {
+                let high_felt = _bytes_to_fixed_recursive(ref self, location + 17, 16, data_len - 16, 0);
+                return u256_from_felts(low_felt, high_felt); 
+            }
```
Let's define just data len conditionally since every other argument is the same

piwonskp commented on warplib/src/warp_memory/bytes.cairo:
```diff
@@ -1,6 +1,51 @@
+    } else {
+        let byte = memory.read(bytes_data_loc);
+        return _bytes_to_fixed_recursive(ref memory, bytes_data_loc + 1, target_width - 1, data_len - 1, 256 * representation + byte);
+    }
```
Imo it'd be more idiomatic to use loop. @rodrigo-pino WYT?

rodrodros committed:
Update workflow

rodrodros committed:
Fix tsc compilation errors

rodrodros committed:
Update warplib testing framwerk

rodrodros committed:
Fix bugs and compilation erros in warplib

rodrodros committed:
Add unit tests to Warplib

AlejandroLabourdette committed:
Refactor and separate type conversions from maths

AlejandroLabourdette committed:
separate external input check functions from math

AlejandroLabourdette committed:
delete autogenerated

AlejandroLabourdette committed:
update gitignore

AlejandroLabourdette committed:
renaming variable

AlejandroLabourdette committed:
Merge branch 'warplib-memory-to-cairo1' into warp-memory-to-cairo1-alab

AlejandroLabourdette committed:
Merge branch 'Update-Warplib-generator' into warp-memory-to-cairo1-alab

rodrodros committed:
Minor bug fixes in accessors

rodrodros committed:
Address review + add new test

rodrodros committed:
Use into

rodrodros committed:
Address review

AlejandroLabourdette committed:
update references to warplib conversions in the branch

AlejandroLabourdette committed:
create directory for autogen if it dosnt exist

AlejandroLabourdette committed:
do not include in git types autogenerated files

AlejandroLabourdette committed:
implementation of fixed bytes types

AlejandroLabourdette committed:
include types in warplib lib

AlejandroLabourdette committed:
try to find traits and structs first to export instead of functions

AlejandroLabourdette committed:
remove support for bitwise operations with fixedBytes

AlejandroLabourdette committed:
Merge branch 'warplib-memory-to-cairo1' into warp-memory-to-cairo1-alab

rodrodros committed:
Update warplib/src/test/warp_memory_test.cairo

Co-authored-by: Piotr Piwoński <piwonskp@gmail.com>

rodrodros committed:
Update warplib/src/test/warp_memory_test.cairo

Co-authored-by: Piotr Piwoński <piwonskp@gmail.com>

rodrodros committed:
Update warplib/src/test/warp_memory_test.cairo

Co-authored-by: Piotr Piwoński <piwonskp@gmail.com>

rodrodros committed:
start_pos -> position

rodrodros committed:
Fix and test change

AlejandroLabourdette committed:
remove bytes handwritten implementation for warpmemory

AlejandroLabourdette committed:
add a generated warp memory bytes implementation

AlejandroLabourdette committed:
add imports needed for warp memory bytes

AlejandroLabourdette committed:
Merge branch 'warplib-memory-to-cairo1' into warp-memory-to-cairo1-alab

AlejandroLabourdette committed:
test for warp memory bytes converter

AlejandroLabourdette committed:
definition for `atIndex` function

AlejandroLabourdette committed:
remove pow2

AlejandroLabourdette committed:
pow2 autogenerated by warplib

AlejandroLabourdette committed:
extend pow2 to n=255

AlejandroLabourdette committed:
pow2 return now u128

AlejandroLabourdette committed:
implement atIndex using pow2

AlejandroLabourdette committed:
tests for atIndex

AlejandroLabourdette committed:
implement bitwise operators for fixed bytes

AlejandroLabourdette committed:
tests for bitwise operations with fixed bytes

AlejandroLabourdette committed:
lint

piwonskp commented:
The PR include commits from @rodrigo-pino branch #1038

AlejandroLabourdette committed:
Merge branch 'cairo-1.0' into warp-memory-to-cairo1-alab

AlejandroLabourdette commented:
@rodrigo-pino could you please take a look at the merge, being more specific: to the warp-memory-test.cairo changes. I give priority to incoming changes in sections that were not addressed by this pr, but seems logic to include some of the former approach.

piwonskp commented on warplib/src/test/fixed_bytes_test.cairo:
```diff
@@ -0,0 +1,115 @@
+    let xor_r: bytes32 = fixed_byte32_1.exclusive_or_bitwise(fixed_byte32_2);
+    let expected_value_xor = u256 { low: 72452722266110920930709281178004808219_u128, high: 72452722266110920930709281178004808219_u128 };
+    assert(xor_r.value == expected_value_xor, 'Unexpected bytes32 bitwise xor');
+}
```
```suggestion
}

```

piwonskp commented on warplib/src/test/warp_memory_test.cairo:
```diff
@@ -307,3 +312,206 @@ fn test_store_retrieve() {
+
+    let byte12 = fixed_byte13.atIndex(12_u8);
+    assert(byte12.value == 199_u8, 'Unexpected byte 12');
+}
```
```suggestion
}

```

AlejandroLabourdette committed:
new line at files end

temyurchenko commented:
@AlejandroLabourdette could you please rebase on cairo-1.0, it seems that many commits are from the already merged branch didn't disappear

AlejandroLabourdette committed:
Merge branch 'cairo-1.0' into warp-memory-to-cairo1-alab

temyurchenko commented:
Unfortunately, merges only add more commits( You'll have to rebase

temyurchenko committed:
fix various issues converted with conversions to the address type (#1060)

- conversions from int variables used to call a Cairo 0 function. It
   was replaced with a new, Cairo 1,
   function---`unsafe_contract_address_from_felt`.

- conversions from int literals were represented in Cairo as raw int
   literals, which fails the type system.

- conversions from int literals to address wouldn't trigger typeString
   change from 'int_const bla-bla' to 'address', which caused
   primitiveTypeToCairo to fail.

- added tests

- the regex had to be changed, as it accidentally was catching methods
   of traits. Anyway, we might need to consider a solution not
   involving regexes, although it's not that bad, given that we only
   impose it on the code written by us

AlejandroLabourdette committed:
rename byte location variable

AlejandroLabourdette committed:
Behavior tests and CI (#1055)

* enable behaviour tests in CI, again

* comments

piwonskp commented on src/warplib/implementations/types/bytes.ts:
```diff
@@ -0,0 +1,168 @@
+      fn exclusive_or_bitwise(self: T, other: T) -> T; // a ^ b
+      // Bitwise negation is not currently supported
+      // Shift operators are not currently supported
+    }
```
Can we store the trait in warplib statically?

AlejandroLabourdette replied:
to do so we would need to create a separate file just for that, then we would have 2 files every time we would want to check something about bytes. Also seems better to have trait definition next to its implementations due that this will not be a commonly extended trait

piwonskp commented on .starknet_accounts_devnet/starknet_open_zeppelin_accounts.json:
Please revert the changes to the file

AlejandroLabourdette committed:
remove starknet accounts json

rodrodros committed:
new warp-plugin bin with bug fixes (#1078)

* Add depured darwin version of warp-plugin

* Add x64 warp plugin binaries

---------

Co-authored-by: Piotr Piwoński <piwonskp>

piwonskp committed:
Do not include starknet accounts file in commits (#1077)

Co-authored-by: Piotr Piwoński <piwonskp>

rodrodros committed:
Update warp memory functionality to Cairo 1 (#1045)

* Make warplib a scarb project

* Add maths module

* update corelib to alpha 7

* Add arm binaries from alpha 7

* Add warp plugin binaries

* Rework structure for

* Fix warp dependency

* Fix warplib and delete files to ignore

* delete autogenerted warplib scripts

* Add corelib and bin for linux_x64

* Scarb package name replaces - for _

* Update compileCairo1 to use new binary

* Update compilation testing framework

* Some comments

* Address review

* Add unimplemented traits for WarpMemory

* Typo

* Delete maths.cairo

* Fix comp errors

* Add array handling logic

* Fix some compilation errors

* Make memory operations safe

* Update part of memory systems

* Add implicit conversion cases

* fix typo

* Array concat

* Minor refactoring to importPath.ts

* Use Serde trait to store/retrieve from memory (#1040)

* Use Serde trait to store/retrieve from memory

* Fix missing import

* Fix function naming in the trait

* Replace individual memory reads with a call to warp_memory

* Divide MemoryTrait into low level trait and external interface

* Remove unused import

* Add missing newline

* Fix wrong imports

* Cell accessor -> Warp memory impl

* Fix missing imports

* Use memory's own `read`instead of Felt252Dict's method

* Add memory implicit to memoryRead

* Fix arrayLiteral implicits

---------

Co-authored-by: Piotr Piwoński <piwonskp>

* Address review

* Update read_id to get_or_create_id

* Remove warplib/target

* Update imports

* Rewrite memory to calldata

* memory to storage

* Update workflow

* Fix tsc compilation errors

* Update warplib testing framwerk

* Fix bugs and compilation erros in warplib

* Add unit tests to Warplib

* Remove warplib sierra

* calldata to memory

* Add storage to memory

* Fix tsc compile bugs

* Add cairo1 compile and format for when using warp

* Update abi imports

* Improvements to memory util func gen

* Other import updates

* Fix some issues with imports

* Update a maths util function

* Minor bug fixes in accessors

* Address review + add new test

* Use into

* Address review

* Update warplib/src/test/warp_memory_test.cairo

Co-authored-by: Piotr Piwoński <piwonskp@gmail.com>

* Update warplib/src/test/warp_memory_test.cairo

Co-authored-by: Piotr Piwoński <piwonskp@gmail.com>

* Update warplib/src/test/warp_memory_test.cairo

Co-authored-by: Piotr Piwoński <piwonskp@gmail.com>

* start_pos -> position

* Fix and test change

* Update warplib/src/test/warp_memory_test.cairo

Co-authored-by: Piotr Piwoński <piwonskp@gmail.com>

* Update warplib/src/test/warp_memory_test.cairo

Co-authored-by: Piotr Piwoński <piwonskp@gmail.com>

* Update warplib/src/test/warp_memory_test.cairo

Co-authored-by: Piotr Piwoński <piwonskp@gmail.com>

* Update warplib/src/test/warp_memory_test.cairo

Co-authored-by: Piotr Piwoński <piwonskp@gmail.com>

* Update warplib/src/test/warp_memory_test.cairo

Co-authored-by: Piotr Piwoński <piwonskp@gmail.com>

* Update warplib/src/test/warp_memory_test.cairo

Co-authored-by: Piotr Piwoński <piwonskp@gmail.com>

* Minor stuff

* Update warplib/src/test/warp_memory_test.cairo

Co-authored-by: Piotr Piwoński <piwonskp@gmail.com>

* Remove starknet_accounts_devnet

* Add super to importPaths

* Fix testing

* Improvments + linting

* Remove mut

* Some minor fixes

* Fix

* linting

* Address review

* Add endent in calldataToMemory

* Improvements in concat util func gen

* Add utilitie

* General fixes

* Improvements

* Update src/cairoUtilFuncGen/memory/implicitConversion.ts

Co-authored-by: Piotr Piwoński <piwonskp@gmail.com>

* Linting

* General fixes and improvements

* Relocation of warplib function

* Update src/cairoUtilFuncGen/memory/memoryStruct.ts

Co-authored-by: Piotr Piwoński <piwonskp@gmail.com>

* Bug fix

* Update src/cairoUtilFuncGen/memory/memoryToStorage.ts

Co-authored-by: Piotr Piwoński <piwonskp@gmail.com>

* Update src/cairoUtilFuncGen/memory/memoryToStorage.ts

Co-authored-by: Piotr Piwoński <piwonskp@gmail.com>

* Update src/cairoUtilFuncGen/memory/memoryToStorage.ts

Co-authored-by: Piotr Piwoński <piwonskp@gmail.com>

* Update src/cairoUtilFuncGen/memory/memoryToCalldata.ts

Co-authored-by: Piotr Piwoński <piwonskp@gmail.com>

* Rename to try

* Endent

* try rename

* Add felt constant and minor improvements

* Typo

* Linting

* Restore .starknet_account_devnet

* Write variables as mut

* Remove triple sss

* Remainig triple sss

* Nitpicks

* remove .starknet_accounts_devnet

* Add some comments

* Untry

* Restore behaviour file to default

* Restore behaviour file to default 2

---------

Co-authored-by: Piotr Piwoński <piwonskp@gmail.com>

cicr99 commented on src/warplib/implementations/types/bytes.ts:
```diff
@@ -0,0 +1,168 @@
+        }
+        fn or_bitwise(self: bytes${length}, other: bytes${length}) -> bytes${length} {
+          ${
+            max_width >= 128 // uN types lower than 128 bits don't have an implementation for bitwise operations
+              ? `bytes${length}{ value: self.value | other.value }`
+              : endent`
+              let self_value_felt252 = u${max_width}_to_felt252(self.value);
+              let other_value_felt252 = u${max_width}_to_felt252(other.value);
+              let self_value_u128 = u128_try_from_felt252(self_value_felt252).unwrap();
+              let other_value_u128 = u128_try_from_felt252(other_value_felt252).unwrap();
+              let result_u128 = self_value_u128 | other_value_u128;
+              let result_felt252 = u128_to_felt252(result_u128);
+              bytes${length}{ value: u${max_width}_try_from_felt252(result_felt252).unwrap() }
+            `
```
This code is the same for add, or and xor right? The only difference is the operator. I think you can write this body inside of a function that also receives the symbol of the operator and returns the string to be written in the cairo output

AlejandroLabourdette committed:
refactor body of bytes bitwise operations

AlejandroLabourdette committed:
delete bytes autogen file

AlejandroLabourdette committed:
accesssor -> accessor

AlejandroLabourdette committed:
Merge branch 'cairo-1.0' into warp-memory-to-cairo1-alab
